ctapipe/io/serializer.py

```diff
@@ -0,0 +1,356 @@
+"""
+Serialize ctapipe containers to file
+"""
+
+from astropy.table import Table, Column
+from ctapipe.core import Container
+from ctapipe.core import Component
+from abc import ABC, abstractmethod
+from astropy import log
+from pickle import load
+from pickle import dump
+from pickle import PickleError
+from traitlets import Unicode
+import numpy as np
+import gzip
+
+__all__ = ['FitsSerializer', 'PickleSerializer']
+log.setLevel('INFO')
+
+
+# PICKLE GZIP Implementation
+
+class Serializer(ABC, Component):
+    """
+    Base class for context manager to save Containers to file
+    """
+    def __init__(self, file,  overwrite=False):
+        """
+        Parameters
+        ----------
+        file:  Unicode
+            input or output full path file name
+        overwrite: Bool
+            overwrite outfile file if it already exists
+        """
+        self.file = file
+        self.overwrite = overwrite
+        super().__init__()
+
+    def __enter__(self):
+        log.debug("Serializing on {0}".format(self.file))
+        return self
+
+    def __exit__(self, *args):
+        pass
+
+    @abstractmethod
+    def add_container(self, container):
+        """
+        Add a container to serializer
+        """
+        pass
+
+    @abstractmethod
+    def write(self):
+        """
+        Write all containers to outfile
+        """
+        pass
+
+    @abstractmethod
+    def __iter__(self):
```

Comment:
I mixed serialization and deserialization in order to be close to Pickle implementation.
There is no sources in io (at least in mater branch)
For your last point, add_container and iteration could not be excecuted on same object because file must be open with a read OR a write flag. An opening with write flag do not allows to read.

_In [4]: file_object = open('test.pickle', 'wb')
In [5]: print(load(file_object))
---------------------------------------------------------------------------
UnsupportedOperation                      Traceback (most recent call last)
<ipython-input-5-c9947b36a037> in <module>()
----> 1 print(load(file_object))

UnsupportedOperation: read
_